File: 03a.pl
#!/usr/bin/perl
use warnings;
use strict;
use feature qw{ say };
use experimental qw{ signatures };

use ARGV::OrDATA;

sub adjacent($y, $from, $to, $schematic) {
    my @adj_y = grep $_ >= 0 && $_ <= $#{ $schematic }, $y - 1, $y, $y + 1;
    my @adj_x = grep $_ >= 0 && $_ <= $#{ $schematic->[$y] },
                $from - 1 .. $to + 1;
    my $symbol = 0;
    for my $j (@adj_y) {
        for my $i (@adj_x) {
            return 1 if $schematic->[$j][$i] =~ /[^0-9.]/;
        }
    }
    return 0
}

my $sum = 0;
my @schematic = map { chomp; [split //] } <>;
for my $y (0 .. $#schematic) {
    my $x = 0;
    while ($x <= $#schematic) {
        if ($schematic[$y][$x] =~ /[0-9]/) {
            my $from = $x;
            ++$x while $x <= $#{ $schematic[$y] }
                       && $schematic[$y][$x] =~ /[0-9]/;
            if (adjacent($y, $from, $x - 1, \@schematic)) {
                my $number = join "", @{ $schematic[$y] }[$from .. $x - 1];
                $sum += $number;
            }
        } else {
            ++$x;
        }
    }
}

say $sum;

__DATA__
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..

File: 03a.py
#! /usr/bin/python3
import sys

def adjacent(n, y, symbol):
    for j in y - 1, y, y + 1:
        if j in symbol:
            for x in symbol[j]:
                if x <= n[1] and x >= n[0]:
                    return True
    return False

DIGITS = {*map(str, range(10))}

number={}
symbol={}
with open(sys.argv[1], 'r') as fh:
    y = 0
    for line in fh:
        line = line.rstrip('\n')
        number[y] = []
        symbol[y] = []
        x = 0
        while x < len(line):
            if line[x] not in DIGITS | {'.'}:
                symbol[y] += [x]
            elif line[x] != '.':
                to = x + 1
                while to < len(line) and line[to] in DIGITS:
                    to += 1
                number[y] += [(x - 1, to, int(line[x:to]))]
                x = to - 1
            x += 1
        y += 1

sum = 0
for y in number:
    for n in number[y]:
        if adjacent(n, y, symbol):
            sum += n[2]
print(sum)
